File: web/js/model/RetirementParameters.js
/**
 * Class representing retirement calculation parameters.
 * This is a JavaScript port of the Java RetirementParameters class.
 */
class RetirementParameters {
    /**
     * Constructor for basic mode parameters
     * 
     * @param {number} annualExpense - Annual expenses in retirement
     * @param {number} retirementPeriod - Number of years in retirement
     * @param {number} expectedReturn - Expected annual return (as decimal, e.g. 0.07 for 7%)
     * @param {number} standardDeviation - Annual standard deviation (as decimal, e.g. 0.10 for 10%)
     * @param {number} inflation - Annual inflation rate (as decimal, e.g. 0.03 for 3%)
     * @param {boolean} adjustForInflation - Whether to adjust expenses for inflation
     * @param {number} numSimulations - Number of Monte Carlo simulations to run
     */
    constructor(annualExpense, retirementPeriod, expectedReturn, standardDeviation, 
                inflation, adjustForInflation, numSimulations) {
        
        // Basic mode parameters
        this.annualExpense = annualExpense;
        this.retirementPeriod = retirementPeriod;
        this.expectedReturn = expectedReturn;
        this.standardDeviation = standardDeviation;
        this.inflation = inflation;
        this.adjustForInflation = adjustForInflation;
        this.numSimulations = numSimulations;
        
        // Default values for advanced mode parameters
        this.currentAge = 0;
        this.retirementAge = 0;
        this.lifeExpectancy = 0;
        this.currentCorpus = 0;
        this.annualContribution = 0;
        this.additionalRetirementIncome = 0;
        this.targetSuccessRate = 85;
    }
    
    /**
     * Set advanced mode parameters
     * 
     * @param {number} currentAge - Current age of the individual
     * @param {number} retirementAge - Target retirement age
     * @param {number} lifeExpectancy - Life expectancy
     * @param {number} currentCorpus - Current retirement corpus
     * @param {number} annualContribution - Annual contribution to retirement corpus
     * @param {number} additionalRetirementIncome - Additional income in retirement (e.g., pension)
     * @param {number} targetSuccessRate - Target success rate (percentage)
     */
    setAdvancedParameters(currentAge, retirementAge, lifeExpectancy, currentCorpus,
                         annualContribution, additionalRetirementIncome, targetSuccessRate) {
        this.currentAge = currentAge;
        this.retirementAge = retirementAge;
        this.lifeExpectancy = lifeExpectancy;
        this.currentCorpus = currentCorpus;
        this.annualContribution = annualContribution;
        this.additionalRetirementIncome = additionalRetirementIncome;
        this.targetSuccessRate = targetSuccessRate;
    }
    
    /**
     * Get the number of years until retirement
     * @returns {number} Years to retirement
     */
    getYearsToRetirement() {
        return this.retirementAge - this.currentAge;
    }
    
    /**
     * Get the number of years in retirement
     * @returns {number} Years in retirement
     */
    getRetirementYears() {
        return this.lifeExpectancy - this.retirementAge;
    }
    
    /**
     * Create a RetirementParameters object from basic mode form data
     * 
     * @param {HTMLFormElement} form - The basic mode form element
     * @returns {RetirementParameters} A new RetirementParameters object
     */
    static fromBasicForm(form) {
        const annualExpense = parseFloat(form.querySelector('#annualExpense').value);
        const retirementPeriod = parseInt(form.querySelector('#retirementPeriod').value);
        const expectedReturn = parseFloat(form.querySelector('#expectedReturn').value) / 100;
        const stdDev = parseFloat(form.querySelector('#stdDev').value) / 100;
        const inflation = parseFloat(form.querySelector('#inflation').value) / 100;
        const adjustForInflation = form.querySelector('#adjustForInflation').checked;
        const numSimulations = parseInt(form.querySelector('#numSimulations').value);
        
        return new RetirementParameters(
            annualExpense,
            retirementPeriod,
            expectedReturn,
            stdDev,
            inflation,
            adjustForInflation,
            numSimulations
        );
    }
    
    /**
     * Create a RetirementParameters object from advanced mode form data
     * 
     * @param {HTMLFormElement} form - The advanced mode form element
     * @returns {RetirementParameters} A new RetirementParameters object
     */
    static fromAdvancedForm(form) {
        const annualExpense = parseFloat(form.querySelector('#advAnnualExpense').value);
        const expectedReturn = parseFloat(form.querySelector('#advExpectedReturn').value) / 100;
        const stdDev = parseFloat(form.querySelector('#advStdDev').value) / 100;
        const inflation = parseFloat(form.querySelector('#advInflation').value) / 100;
        const numSimulations = parseInt(form.querySelector('#advNumSimulations').value);
        
        // Create basic parameters
        const params = new RetirementParameters(
            annualExpense,
            0, // Will be calculated from lifespans
            expectedReturn,
            stdDev,
            inflation,
            true, // Always adjust for inflation in advanced mode
            numSimulations
        );
        
        // Add advanced parameters
        const currentAge = parseInt(form.querySelector('#currentAge').value);
        const retirementAge = parseInt(form.querySelector('#retirementAge').value);
        const lifeExpectancy = parseInt(form.querySelector('#lifeExpectancy').value);
        const currentCorpus = parseFloat(form.querySelector('#currentCorpus').value);
        const annualContribution = parseFloat(form.querySelector('#annualContribution').value);
        const additionalIncome = parseFloat(form.querySelector('#additionalIncome').value);
        const targetSuccessRate = parseFloat(form.querySelector('#targetSuccessRate').value);
        
        params.setAdvancedParameters(
            currentAge,
            retirementAge,
            lifeExpectancy,
            currentCorpus,
            annualContribution,
            additionalIncome,
            targetSuccessRate
        );
        
        return params;
    }
}

File: src/main/java/com/retirement/model/RetirementParameters.java
package com.retirement.model;

/**
 * Model class to store retirement parameters for calculations.
 */
public class RetirementParameters {
    // Basic mode parameters
    private double annualExpense;
    private int retirementPeriod;
    private double expectedReturn;
    private double standardDeviation;
    private double inflation;
    private boolean adjustForInflation;
    
    // Advanced mode parameters
    private int currentAge;
    private int retirementAge;
    private int lifeExpectancy;
    private double currentCorpus;
    private double annualContribution;
    private double additionalRetirementIncome;
    private double targetSuccessRate;
    
    // Simulation parameters
    private int numSimulations;
    
    // Constructor for basic mode
    public RetirementParameters(double annualExpense, int retirementPeriod, 
                               double expectedReturn, double standardDeviation, 
                               double inflation, boolean adjustForInflation,
                               int numSimulations) {
        this.annualExpense = annualExpense;
        this.retirementPeriod = retirementPeriod;
        this.expectedReturn = expectedReturn;
        this.standardDeviation = standardDeviation;
        this.inflation = inflation;
        this.adjustForInflation = adjustForInflation;
        this.numSimulations = numSimulations;
    }
    
    // Constructor for advanced mode
    public RetirementParameters(int currentAge, int retirementAge, int lifeExpectancy,
                               double currentCorpus, double annualExpense, 
                               double annualContribution, double additionalRetirementIncome,
                               double expectedReturn, double standardDeviation, 
                               double inflation, double targetSuccessRate,
                               int numSimulations) {
        this.currentAge = currentAge;
        this.retirementAge = retirementAge;
        this.lifeExpectancy = lifeExpectancy;
        this.currentCorpus = currentCorpus;
        this.annualExpense = annualExpense;
        this.annualContribution = annualContribution;
        this.additionalRetirementIncome = additionalRetirementIncome;
        this.expectedReturn = expectedReturn;
        this.standardDeviation = standardDeviation;
        this.inflation = inflation;
        this.targetSuccessRate = targetSuccessRate;
        this.numSimulations = numSimulations;
        this.adjustForInflation = true; // Always adjust for inflation in advanced mode
    }
    
    // Getters and setters
    public double getAnnualExpense() {
        return annualExpense;
    }
    
    public void setAnnualExpense(double annualExpense) {
        this.annualExpense = annualExpense;
    }
    
    public int getRetirementPeriod() {
        return retirementPeriod;
    }
    
    public void setRetirementPeriod(int retirementPeriod) {
        this.retirementPeriod = retirementPeriod;
    }
    
    public double getExpectedReturn() {
        return expectedReturn;
    }
    
    public void setExpectedReturn(double expectedReturn) {
        this.expectedReturn = expectedReturn;
    }
    
    public double getStandardDeviation() {
        return standardDeviation;
    }
    
    public void setStandardDeviation(double standardDeviation) {
        this.standardDeviation = standardDeviation;
    }
    
    public double getInflation() {
        return inflation;
    }
    
    public void setInflation(double inflation) {
        this.inflation = inflation;
    }
    
    public boolean isAdjustForInflation() {
        return adjustForInflation;
    }
    
    public void setAdjustForInflation(boolean adjustForInflation) {
        this.adjustForInflation = adjustForInflation;
    }
    
    public int getCurrentAge() {
        return currentAge;
    }
    
    public void setCurrentAge(int currentAge) {
        this.currentAge = currentAge;
    }
    
    public int getRetirementAge() {
        return retirementAge;
    }
    
    public void setRetirementAge(int retirementAge) {
        this.retirementAge = retirementAge;
    }
    
    public int getLifeExpectancy() {
        return lifeExpectancy;
    }
    
    public void setLifeExpectancy(int lifeExpectancy) {
        this.lifeExpectancy = lifeExpectancy;
    }
    
    public double getCurrentCorpus() {
        return currentCorpus;
    }
    
    public void setCurrentCorpus(double currentCorpus) {
        this.currentCorpus = currentCorpus;
    }
    
    public double getAnnualContribution() {
        return annualContribution;
    }
    
    public void setAnnualContribution(double annualContribution) {
        this.annualContribution = annualContribution;
    }
    
    public double getAdditionalRetirementIncome() {
        return additionalRetirementIncome;
    }
    
    public void setAdditionalRetirementIncome(double additionalRetirementIncome) {
        this.additionalRetirementIncome = additionalRetirementIncome;
    }
    
    public double getTargetSuccessRate() {
        return targetSuccessRate;
    }
    
    public void setTargetSuccessRate(double targetSuccessRate) {
        this.targetSuccessRate = targetSuccessRate;
    }
    
    public int getNumSimulations() {
        return numSimulations;
    }
    
    public void setNumSimulations(int numSimulations) {
        this.numSimulations = numSimulations;
    }
    
    /**
     * Calculate the total retirement years from retirement age to life expectancy.
     * @return Total retirement years
     */
    public int getRetirementYears() {
        return lifeExpectancy - retirementAge;
    }
    
    /**
     * Calculate the years until retirement from current age.
     * @return Years until retirement
     */
    public int getYearsToRetirement() {
        return retirementAge - currentAge;
    }
}
